[core] Layerwise Upcasting (#10347)
huggingface/diffusers: docs/source/en/optimization/memory.md

</Tip>

## FP8 layerwise weight-casting

> PyTorch supports `torch.float8_e4m3fn` and `torch.float8_e5m2` as weight storage dtypes. This precision cannot be used for performing computation directly for many different tensor operations due to unimplemented kernel support. However, one can still use these dtypes for storing model weights in lower FP8 precision. For computation, the weights can be upcasted on-the-fly as and when layers are invoked in the forward pass.

Reviewer: ```suggestion
PyTorch supports `torch.float8_e4m3fn` and `torch.float8_e5m2` as weight storage dtypes, but they can't be used for computation in many different tensor operations due to unimplemented kernel support. However, you can use these dtypes to store model weights in fp8 precision and upcast them on-the-fly when the layers are used in the forward pass. This is known as layerwise weight-casting.
```

> Typically, inference on most models is done with `torch.float16` or `torch.bfloat16` weight/computation precision. Applying layerwise weight-casting, by storing the weights in FP8 precision, cuts down the memory footprint of the model weights by half approximately.

Reviewer: ```suggestion
Typically, inference on most models is done with `torch.float16` or `torch.bfloat16` weight/computation precision. Layerwise weight-casting cuts down the memory footprint of the model weights by approximately half.
```

```python
import torch
from diffusers import CogVideoXPipeline
from diffusers.utils import export_to_video

> pipe = CogVideoXPipeline.from_pretrained("THUDM/CogVideoX-5b", torch_dtype=torch.bfloat16)

Author: Thanks, missed it! I've updated the example.

I think we should also try to support directly loading in `float8_*` types in `from_pretrained` in a future PR, which respects `_keep_modules_in_fp32` and `_always_upcast_modules` (I'm not too sure if this is a good variable name btw so open to suggestions)

> pipe = CogVideoXPipeline.from_pretrained("THUDM/CogVideoX-5b", torch_dtype=torch.bfloat16)

Author: I've renamed this to `_precision_sensitive_module_patterns` but not really sure on the name. Would love to iterate again if someone has suggestions

> pipe = CogVideoXPipeline.from_pretrained("THUDM/CogVideoX-5b", torch_dtype=torch.bfloat16)

Reviewer 2: Nice name.

pipe.to("cuda")
pipe.transformer.enable_layerwise_upcasting(storage_dtype=torch.float8_e4m3fn, compute_dtype=torch.bfloat16)

prompt = (
    "A panda, dressed in a small, red jacket and a tiny hat, sits on a wooden stool in a serene bamboo forest. "
    "The panda's fluffy paws strum a miniature acoustic guitar, producing soft, melodic tunes. Nearby, a few other "
    "pandas gather, watching curiously and some clapping in rhythm. Sunlight filters through the tall bamboo, "
    "casting a gentle glow on the scene. The panda's face is expressive, showing concentration and joy as it plays. "
    "The background includes a small, flowing stream and vibrant green foliage, enhancing the peaceful and magical "
    "atmosphere of this unique musical performance."
)
video = pipe(prompt=prompt, guidance_scale=6, num_inference_steps=50).frames[0]
export_to_video(video, "output.mp4", fps=8)
```

> In the above example, we enable layerwise upcasting on the transformer component of the pipeline. By default, certain layers are skipped from the FP8 weight casting because it can lead to significant degradation of generation quality. For most cases, skipping the normalization and modulation related weight parameters is a good choice (which is also the default choice). However, more control and flexibility can be obtained by directly utilizing the [`~hooks.layerwise_upcasting.apply_layerwise_upcasting`] function instead of using [`~ModelMixin.enable_layerwise_upcasting`].

Reviewer: I think maybe a sentence or two explaining why `apply_layerwise_upcasting` is more flexible would be helpful/easier for users to connect the dots + maybe even show a code example of it.

```suggestion
In the above example, layerwise upcasting is enabled on the transformer component of the pipeline. By default, certain layers are skipped from the FP8 weight casting because it can lead to significant degradation of generation quality. The normalization and modulation related weight parameters are also skipped by default.

However, you gain more control and flexibility by directly utilizing the [`~hooks.layerwise_upcasting.apply_layerwise_upcasting`] function instead of [`~ModelMixin.enable_layerwise_upcasting`].
```

> [[autodoc]] ModelMixin.enable_layerwise_upcasting

Reviewer: I'd add this to the API reference [here](https://huggingface.co/docs/diffusers/api/models/overview)

Author: Thanks @stevhliu! Updated

> [[autodoc]] hooks.layerwise_upcasting.apply_layerwise_upcasting

Reviewer: This one can maybe be moved [here](https://huggingface.co/docs/diffusers/api/utilities)?
